File: python/MinStack.py
class MinStack(object):
    l=[]
    m=2147483648
    def __init__(self):
        """
        initialize your data structure here.
        """
        self.l=[]
        

    def push(self, x):
        """
        :type x: int
        :rtype: void
        """
        if x<self.m:
            self.m=x
        self.l.append(x)
        

    def pop(self):
        """
        :rtype: void
        """
        if len(self.l)==1:
            self.m=2147483648
        elif self.m==self.l[len(self.l)-1]:
            self.m=min(self.l[0:len(self.l)-1])
        del self.l[len(self.l)-1]
        

    def top(self):
        """
        :rtype: int
        """
        return self.l[len(self.l)-1]
        

    def getMin(self):
        """
        :rtype: int
        """
        return self.m
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: C/MinStack.c
#include <limits.h>
#include <stdlib.h>

#define min(a, b) (((a) < (b)) ? (a) : (b))

typedef struct
{
	int stack[10000][2];  //{value,current min}
	int top;
} MinStack;

/** initialize your data structure here. */

MinStack* minStackCreate()
{
	MinStack* stack = (MinStack*)malloc(sizeof(MinStack));
	stack->top = 0;
	stack->stack[0][0] = 0;
	stack->stack[0][1] = INT_MAX;
	return stack;
}

void minStackPush(MinStack* obj, int x)
{
	obj->stack[obj->top + 1][1] = min(x, obj->stack[obj->top][1]);
	obj->stack[++obj->top][0] = x;
}

void minStackPop(MinStack* obj)
{
	--obj->top;
}

int minStackTop(MinStack* obj)
{
	return obj->stack[obj->top][0];
}

int minStackGetMin(MinStack* obj)
{
	return obj->stack[obj->top][1];
}

void minStackFree(MinStack* obj)
{
	if (obj)
		free(obj);
}

/**
 * Your MinStack struct will be instantiated and called as such:
 * MinStack* obj = minStackCreate();
 * minStackPush(obj, x);
 
 * minStackPop(obj);
 
 * int param_3 = minStackTop(obj);
 
 * int param_4 = minStackGetMin(obj);
 
 * minStackFree(obj);
*/
